python/cudf/cudf/core/buffer/spill_manager.py

```diff
@@ -0,0 +1,363 @@
+
+        # Print spilling stats
+        ret += " Spilling (level >= 1):"
+        if len(self.spills_totals) == 0:
+            ret += " None"
+        ret += "\n"
+        for (src, dst), (nbytes, time) in self.spills_totals.items():
+            ret += f"  {src} => {dst}: {format_bytes(nbytes)} in {time}s\n"
+
+        # Print expose stats
+        ret += " Expose (level >= 2):"
+        if self.level < 2:
+            return ret + " disabled"
+        if len(self.expose) == 0:
+            ret += " None"
+        ret += "\n"
+        for s in sorted(self.expose.values(), key=lambda x: -x.count):
+            ret += (
+                f" Count: {s.count}, total: {format_bytes(s.total_nbytes)}, "
+            )
+            ret += f"spilled: {format_bytes(s.spilled_nbytes)}\n"
+            ret += s.traceback
+            ret += "\n"
+        return ret[:-1]
+
+    def log_spill(self, src: str, dst: str, nbytes: int, time: float) -> None:
+        if self.level < 1:
+            return
+        with self.lock:
+            total_nbytes, total_time = self.spills_totals[(src, dst)]
+            self.spills_totals[(src, dst)] = (
+                total_nbytes + nbytes,
+                total_time + time,
+            )
+
+    def log_expose(self, buf: SpillableBuffer) -> None:
+        if self.level < 2:
+            return
+
+        with self.lock:
+            tb = get_traceback()
+            stat = self.expose.get(tb, None)
+            spilled_nbytes = buf.nbytes if buf.is_spilled else 0
+            if stat is None:
+                self.expose[tb] = ExposeStatistic(
+                    traceback=tb,
+                    total_nbytes=buf.nbytes,
+                    spilled_nbytes=spilled_nbytes,
+                )
+            else:
+                stat.count += 1
+                stat.total_nbytes += buf.nbytes
+                stat.spilled_nbytes += spilled_nbytes
+
+
+class SpillManager:
+    """Manager of spillable buffers.
+
+    This class implements tracking of all known spillable buffers, on-demand
+    spilling of said buffers, and (optional) maintains a memory usage limit.
+
+    When `spill_on_demand=True`, the manager registers an RMM out-of-memory
+    error handler, which will spill spillable buffers in order to free up
+    memory.
+
+    When `device_memory_limit=True`, the manager will try keep the device
+    memory usage below the specified limit by spilling of spillable buffers
+    continuously, which will introduce a modest overhead.
+
+    Parameters
+    ----------
+    spill_on_demand : bool
+        Enable spill on demand. The global manager sets this to the value of
+        `CUDF_SPILL_ON_DEMAND` or False.
+    device_memory_limit: int, optional
+        If not None, this is the device memory limit in bytes that triggers
+        device to host spilling. The global manager sets this to the value
+        of `CUDF_SPILL_DEVICE_LIMIT` or None.
+    """
+
+    _base_buffers: MutableMapping[int, SpillableBuffer]
```

Comment:
Can we just type this more precisely as a WeakValueDictionary?

More importantly, why is it a dict rather than just a WeakList? Why do we need to keep a key-value mapping when the keys are arbitrary counters?

Reply:
> Can we just type this more precisely as a WeakValueDictionary?

Done

> More importantly, why is it a dict rather than just a WeakList? Why do we need to keep a key-value mapping when the keys are arbitrary counters?

We don't but AFAIK `weakref` only implements weak `dict` and `set`?